File: matlab/lms_circular.m
ABSPATH = pwd;
% identify current folder (absolute path)
addpath([ABSPATH '/function']);
% add function path

load('lms_data.mat');

LMS_LENGTH = 300;
STEP_SIZE = 0.04;
TOTAL_LENGTH = length(cmd_noise);

weights = zeros(1, LMS_LENGTH);
noise_buffer = zeros(LMS_LENGTH, 1);

err = zeros(TOTAL_LENGTH, 1);

norm_factor = 0;

for n = 1:LMS_LENGTH
	output = 0;
	for k = 1:n
		output = output + weights(k) * noise(n-k+1);
	end
	for k = n+1:LMS_LENGTH
		output = output + weights(k) * noise_buffer(n-k+LMS_LENGTH+1);
	end

	err(n) = cmd_noise(n) - output;

	norm_factor = norm_factor - noise_buffer(n)^2 + noise(n)^2;
	new_step_size = STEP_SIZE / (1+norm_factor);

	for k = 1:n
		weights(k) = weights(k) + new_step_size * err(n) * noise(n-k+1);
	end
	for k = n+1:LMS_LENGTH
		weights(k) = weights(k) + new_step_size * err(n) * noise_buffer(n-k+LMS_LENGTH+1);
	end
end

for n = (LMS_LENGTH+1):TOTAL_LENGTH
	output = 0;
	for k = 1:LMS_LENGTH
		output = output + weights(k) * noise(n-k+1);
	end

	err(n) = cmd_noise(n) - output;

	norm_factor = norm_factor - noise(n-LMS_LENGTH)^2 + noise(n)^2;
	new_step_size = STEP_SIZE / (1+norm_factor);

	for k = 1:LMS_LENGTH
		weights(k) = weights(k) + new_step_size * err(n) * noise(n-k+1);
	end

	for k = 1:LMS_LENGTH
		noise_buffer(k) = noise(TOTAL_LENGTH-LMS_LENGTH+k);
	end
end

File: pc/lms_circular.c
#include <stdio.h>

#include "lms_data.c"

#define TOTAL_LENGTH	48000
#define LMS_LENGTH		300
#define LMS_STEP_SIZE	0.04

double err[TOTAL_LENGTH] = { 0.0 };
double weights[LMS_LENGTH] = { 0.0 };
double noise_buffer[LMS_LENGTH] = { 0.0 };
double square_sum = 0;

int main() {
	double output;
	double new_step_size;

	int n;
	int k;

	for (n = 0; n < LMS_LENGTH; n++) {
		output = 0;
		for (k = 0; k <= n; k++) {
			output += weights[k] * noise[n-k];
		}
		for (k = n+1; k < LMS_LENGTH; k++) {
			output += weights[k] * noise_buffer[n-k+LMS_LENGTH];
		}

		err[n] = cmd_noise[n] - output;

		square_sum = square_sum - noise_buffer[n] * noise_buffer[n] + noise[n] * noise[n];
		new_step_size = LMS_STEP_SIZE / (1+square_sum);

		for (k = 0; k <= n; k++) {
			weights[k] += new_step_size * err[n] * noise[n-k];
		}
		for (k = n+1; k < LMS_LENGTH; k++) {
			weights[k] += new_step_size * err[n] * noise_buffer[n-k+LMS_LENGTH];
		}
	}

	for (n = LMS_LENGTH; n < TOTAL_LENGTH; n++) {
		output = 0;
		for (k = 0; k < LMS_LENGTH; k++) {
			output += weights[k] * noise[n-k];
		}

		err[n] = cmd_noise[n] - output;

		square_sum = square_sum - noise[n-LMS_LENGTH] * noise[n-LMS_LENGTH] + noise[n] * noise[n];
		new_step_size = LMS_STEP_SIZE / (1+square_sum);

		for (k = 0; k < LMS_LENGTH; k++) {
			weights[k] += new_step_size * err[n] * noise[n-k];
		}
	}

	for (n = 0; n < LMS_LENGTH; n++) {
		noise_buffer[n] = noise[TOTAL_LENGTH-LMS_LENGTH+n];
	}

	for (n = 0; n < TOTAL_LENGTH; n++) {
		printf("%e\n", err[n]);
	}

	return 0;
}
